packages/common/src/models/Track.ts

```diff
@@ -276,6 +276,20 @@ export type TrackMetadata = {
 
   offline?: OfflineTrackMetadata
   local?: boolean
+
+  // API returned URLS
+  stream?: {
+    url?: string
```

Comment:
would it ever be possible for url to be `undefined`?

Reply:
Yes. It's a bit annoying:

1. `stream` is undefined on non-full tracks.
2. `stream` is defined always on full tracks, but `url` is _null_ on the actual response JSON when you don't have access/no signature was sent from DN.
3. `url` is _undefined_ in the SDK response because it gets remapped.

Most of our other fields are marked as nullable in this model, so the adapter has to do some coercing back from undefined to null for those already, but @schottra and I thought it made more sense to leave this new addition as optional instead of remapping it back to null.

Reply:
counter bump ok.

probably there exists a cleaner data model rather than trying to fight this. 

if you don't have access it would make sense that you could get a stream url back that would 403 instead of being absent but eh...

Reply:
was tempted to have it be the DN stream url

Reply:
i don't hate it... if we went off of fill i would say that's a reasonable idea haha

Reply:
What did you mean by fill? Did you mean feel?

Reply:
FULL***

Reply:
Yeah just for some context here: It is indeed very hard to get the correct types out of flask. Our SDK generator also doesn't really support them correctly. We talked about this when we started the SDK migration. I went down a path of updating the generator to make optional fields show up as `value?: string | null` so that we wouldn't have to do the adapters in client to go from `value?: string` -> `value?: string | null` since we chose optional AND Nullable for a lot of our types in client side models (maybe this is for edit flows when you want to clear a value?)

Ultimately I think our desired end state is that SDK models render optional fields as just optional, not nullable (currently true) and that client updates to make optional fields just optional, not nullable.

And for updates, we can either leverage separate types with nullable fields for deleting values, or we require that you pass the full object in an update and we treat missing optional fields as deletions.


Regarding `stream` and `url`, two issues:
1. Using the same target type for full and standard tracks on client side makes this a headache. Otherwise we would just say `stream` is always defined but may have no `url`
2. Setting that aside, working from a full model...nested fields should either always be defined and have optional fields beneath them if there is a logical reason (i.e. you don't have access) or they should be optional, but fully formed when specified. It's also a headache to have optional nested objects with optional / nullable fields. So the more we can cut down on this and move to a world where we don't have to run any conversion logic on SDK models to get them in line with client models (other than snake-casing), the better.